clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes.hs

```diff
@@ -0,0 +1,215 @@
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  knownNatN, _knownDomainWrite, _knownDomainRead, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    --  , _, _, _, _, _, _, _
+     ] = lefts args
+
+  dcFifoName <- Id.makeBasic "dcfifo"
+
+  let depth =
+        case knownNatN of
+          Literal (NaturalLiteral n) -> fromInteger n
+          _ -> error "Unexpected type of knownNatN!"
+
+  pure (Right (bbMeta depth dcFifoName dcConfig, bb dcFifoName dcConfig))
+ where
+  bbMeta width dcFifoName dcConfig = emptyBlackBoxMeta
+    { N.bbKind = N.TDecl
+    , N.bbIncludes =
+      [ ( ("dcfifo", "tcl")
+        , BBFunction (show 'dcFifoTclTF) 0 (dcFifoTclTF width dcFifoName dcConfig))
+      ]
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    -- , N.bbResultNames =
+    --   , N.BBTemplate [N.Text "wr_full"]
+    --   , N.BBTemplate [N.Text "wr_data_count"]
+
+    --   , N.BBTemplate [N.Text "rd_empty"]
+    --   , N.BBTemplate [N.Text "rd_data_count"]
+    --   , N.BBTemplate [N.Text "rd_dout"]
+    --   ]
+    }
+
+  bb :: Identifier -> DcConfig TermLiteralSNat -> BlackBox
+  bb dcFifoName dcConfig = BBFunction (show 'dcFifoTF) 0 (dcFifoTF dcFifoName dcConfig)
+
+-- | Instantiate IP generate with 'dcFifoTclTF'.
+dcFifoTF :: Identifier -> DcConfig TermLiteralSNat -> TemplateFunction
+dcFifoTF dcFifoName DcConfig{..} = TemplateFunction [] (const True) $ \bbCtx -> do
+  let
+    TermLiteralSNat (fromIntegral -> depth) = dcDepth
+
+    [  _knownNatN, _knownDomainWrite, _knownDomainRead, _knownNatDepth
+     , _constraint1, _constraint2
+     , _dcConfig
+     , wClk, rClk, rst, wDataM
+     , rEnable
+     ] = map fst (DSL.tInputs bbCtx)
+
+    [tResult] = map DSL.ety (DSL.tResults bbCtx)
+
+  dcFifoInstName <- Id.makeBasic "dcfifo_inst"
+
+  -- -1 for maybe
+  let dataTy = N.BitVector (DSL.tySize (DSL.ety wDataM) - 1)
+  let
+    blockInps =
+      [ ("rst", N.Bit)
+      , ("wr_clk", N.Bit)
+      , ("rd_clk", N.Bit)
+      , ("din", dataTy)
+      , ("wr_en", N.Bit)
+      , ("rd_en", N.Bit)
+      ]
+    blockOuts = catMaybes
+      [ Just ("dout", dataTy)
+      , Just ("full", N.Bit)
+      , Just ("empty", N.Bit)
+      , dcReadDataCount `orNothing` ("rd_data_count", N.BitVector depth)
+      , dcWriteDataCount `orNothing` ("wr_data_count", N.BitVector depth)
+      , dcUnderflow `orNothing` ("underflow", N.Bit)
+      , dcOverflow `orNothing` ("overflow", N.Bit)
+      ]
+
+  DSL.declarationReturn bbCtx "dcfifo_inst_block" $ do
+
+    DSL.compInBlock "dcfifo" blockInps blockOuts
+
+    (wEna, wData) <- DSL.deconstructMaybe wDataM ("wr_ena", "rd_din")
```

Comment:
You define signals named `wr_ena` and `rd_enable`. Could you rename one of those to be consistent?

Also, **more importantly**, I think that should be `wr_din`, not `rd_din`. The prefix indicates the clock domain, right? And you also give it the name `wData`, with a `w`.